### ISSUE TITLE
Can't fix failed allocation

### ISSUE THREAD

**Issue by apollo13** [hashicorp/nomad#18620]
### Nomad version
Nomad v1.6.2
BuildDate 2023-09-13T16:47:25Z
Revision 73e372ad94033db2ceaf53468b270a31544c23fd

### Operating system and Environment details
Debian 11 with docker driver

### Issue
I have a failed deployment where one of the allocations failed to upgrade during deployment since it couldn't download the container fast enough:
![image](https://github.com/hashicorp/nomad/assets/81547/1ea8b392-4a56-49b8-9c53-430e0d53273e)

Since 15:58 it has failed completely, interesting enough it still tries to kill (?) the task (ie see 16:19). Any ideas what is wrong here and is there any command I can use to get that failed allocation back again?

### Reproduction steps
No idea :)

**Comment by apollo13** [comment]
Okay, so I have been playing with all nomad cli commands that I could find and nothing gets that allocation back. For instance `nomad job restart -reschedule -namespace infra elasticsearch` hangs at:
```
==> 2023-09-29T21:59:30+02:00: Restarting 4 allocations
    2023-09-29T21:59:30+02:00: Rescheduling allocation "8c240d7a" for group "elasticsearch"
    2023-09-29T22:00:30+02:00: Still waiting for allocation "8c240d7a" to be replaced
```
I am running out of ideas.

**Comment by apollo13** [comment]
I had to run it in the namespace:
`nomad alloc stop -namespace infra -no-shutdown-delay 8c240d7a`
and it basically immediately returned:
```
==> 2023-10-04T08:33:06+02:00: Monitoring evaluation "806b2989"
    2023-10-04T08:33:06+02:00: Evaluation triggered by job "elasticsearch"
    2023-10-04T08:33:07+02:00: Evaluation within deployment: "be5c6a9e"
    2023-10-04T08:33:07+02:00: Evaluation status changed: "pending" -> "complete"
==> 2023-10-04T08:33:07+02:00: Evaluation "806b2989" finished with status "complete"
==> 2023-10-04T08:33:07+02:00: Monitoring deployment "be5c6a9e"
  ! Deployment "be5c6a9e" failed
    
    2023-10-04T08:33:07+02:00
    ID          = be5c6a9e
    Job ID      = elasticsearch
    Job Version = 16
    Status      = failed
    Description = Failed due to progress deadline
    
    Deployed
    Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
    elasticsearch  3        3       2        1          2023-09-29T13:57:19Z
    kibana         1        1       1        0          2023-09-29T15:45:54+02:00
```

**Comment by apollo13** [comment]
Hi @lgfa29,

I didn't want to leave the cluster in this state any longer so I did the following: I added `DUMMY = 1` as env var to the tasks in question which should forcibly recreate them and ran a plan:
```
+/- Job: "elasticsearch"
+/- Task Group: "elasticsearch" (1 create, 2 ignore)
      Task: "connect-proxy-elasticsearch"
  +/- Task: "elasticsearch" (forces create/destroy update)
    + Env[DUMMY]: "1"

+/- Task Group: "kibana" (1 create/destroy update)
      Task: "connect-proxy-kibana"
  +/- Task: "kibana" (forces create/destroy update)
    + Env[DUMMY]: "1"

Scheduler dry-run:
- All tasks successfully allocated.

Job Modify Index: 2554831
To submit the job with version verification run:

nomad job run -check-index 2554831 elasticsearch.nomad

When running the job with the check-index flag, the job will only be run if the
job modify index given matches the server-side version. If the index has
changed, another user has modified the job and the plan's results are
potentially invalid.
```

Why does it want to create only a single instance of elasticsearch and ignore the other two? This makes no sense for an env variable (though I guess it might simply be the plan output that is broken here?)

Deploying that did actually recreate all instances:
```
==> 2023-10-06T09:17:35+02:00: Monitoring evaluation "1426d89b"
    2023-10-06T09:17:36+02:00: Evaluation triggered by job "elasticsearch"
    2023-10-06T09:17:36+02:00: Evaluation within deployment: "c216c14e"
    2023-10-06T09:17:36+02:00: Allocation "295b9d01" created: node "42eb20db", group "kibana"
    2023-10-06T09:17:36+02:00: Allocation "57f2b1a1" created: node "42eb20db", group "elasticsearch"
    2023-10-06T09:17:36+02:00: Evaluation status changed: "pending" -> "complete"
==> 2023-10-06T09:17:36+02:00: Evaluation "1426d89b" finished with status "complete"
==> 2023-10-06T09:17:36+02:00: Monitoring deployment "c216c14e"
  ✓ Deployment "c216c14e" successful
    
    2023-10-06T09:22:48+02:00
    ID          = c216c14e
    Job ID      = elasticsearch
    Job Version = 17
    Status      = successful
    Description = Deployment completed successfully
    
    Deployed
    Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
    elasticsearch  3        3       3        0          2023-10-06T07:37:46Z
    kibana         1        1       1        0          2023-10-06T09:30:15+02:00
```

**Comment by lgfa29** [comment]
Hi @apollo13, 

Apologies for the delay and thank you for the keeping the environment broken for us.

I think I was able to create a quick reproduction of this problem:

```hcl
job "crash" {
  group "crash" {
    count = 3

    update {
      max_parallel      = 3
      min_healthy_time  = "1s"
      healthy_deadline  = "3s"
      progress_deadline = "5s"
      stagger           = "1s"
    }

    restart {
      attempts = 1
      delay    = "0s"
    }

    task "crash" {
      driver = "docker"

      config {
        image   = "alpine:3"
        command = "/bin/ash"
        args    = ["local/script.sh"]
      }

      template {
        data        = <<EOF
if [ "$NOMAD_ALLOC_INDEX" == "2" ]; then
  exit 1
fi

tail -f /dev/null
EOF
        destination = "local/script.sh"
      }
    }
  }
}
```

The actual `update` and `restart` values are not really relevant, I just reduced them quite a bit to reproduce the problem more quickly. 

Looking at the logs it seems like the Nomad scheduler is completely ignoring the failed allocation:

```
    2023-10-07T00:36:17.365Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=6df6cd97-a382-85c2-0161-7ae16d504ea5 job_id=crash namespace=default worker_id=7f88d19e-98fb-59e9-de82-86cf50811463
  results=
  | Total changes: (place 0) (destructive 0) (inplace 0) (stop 0) (disconnect 0) (reconnect 0)
  | Desired Changes for "crash": (place 0) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 2) (canary 0)
```

That's _extra ignoring_ because it's not even taking into account (the remaining healthy allocs are properly ignored 😄).

This may be related to https://github.com/hashicorp/nomad/issues/1576 but made worse by the fact the allocation failed and Nomad should have tried to replace it.

I will put this issue into our roadmap for further investigation but at least we now have a reproduction so you won't need to keep your deployments broken any longer 😅 

Thank you so much for the help!

**Comment by apollo13** [comment]
Hi @lgfa29, no worries -- great you can reproduce it. Stupid question though, I am not a paying customer, but would there be some way for me to escalate such issues? Not to have them fixed or get priority treatment but to be able to provide you with enough information so they can be fixed in the longrun. After all I think this issue does not only affect me and others (paying customers with prod workloads) might not be in the situation to leave the cluster in a semi-broken state for long.

P.S.: If it helps, you can look at the PRs/issues authored by me and you will see that I try to be a good OSS supporter and contribute back as much as I can and my unpaid time permits. Or my CSI plugin ( https://gitlab.com/rocketduck/csi-plugin-nfs ) which is explicitly written for nomad (or rather written in a way to not have any orchestrator dependency) which is used by quite a few people and from what I heard even internally at hashicorp for testing because it is so easy to setup and use (at least when the hostpath plugin doesn't suffice).

**Comment by lgfa29** [comment]
> I am not a paying customer, but would there be some way for me to escalate such issues? Not to have them fixed or get priority treatment but to be able to provide you with enough information so they can be fixed in the longrun. After all I think this issue does not only affect me and others (paying customers with prod workloads) might not be in the situation to leave the cluster in a semi-broken state for long.

I don't think we have anything like this. I do acknowledge and apologize for the delay in getting back to you in this issue given that you held a broken cluster state to help us debug the problem, but async communication is the way we're able to balance our different work streams.

The work you do for Nomad is amazing, and we really appreciate it. We usually don't encourage excessive pings on GitHub, but from people we know (like yourself), I personally wouldn't mind additional pings in a situation like this. 

Another option to balance live debug data and async communication would be to [capture debug bundles](https://support.hashicorp.com/hc/en-us/articles/4888993669139-How-to-Create-Nomad-Operator-Debug-Bundle) and email them to us at nomad-oss-debug@hashicorp.com referencing the issue number in the subject. Then you unblock your cluster as the bundle contains a lot of runtime information that may help us nail down the problem.

But this is an important question that I unfortunately don't have a good answer for. I have raised this with the team to see if we can come up with something better.

Thank you again for raising this and all the work you do 🙂

**Comment by apollo13** [comment]
Hello Luiz, thank you for the kind words. There is no need to apologize -- I understand that other work is more important and that the bulk of the work happens async with different timezones adding plenty of delay. Be prepared for some pings in the future :) 

I will send stuff to nomad-oss-debug next time then even without someone from the team requesting data before, I usually only did it when someone explicitly requested it.

**Comment by Tyrael** [comment]
it looks doing
`nomad system gc` (to clean up the failed allocs) and then modifying the definition to trigger a deployment can be used as a workaround
